feat(shader-picker): new handling of shaders

NMC-TBone opened the pull request:
closes: #180 #179 #170 #214 #68

## FS22/General implementation
* New UI for shader selection and parameter management
* Added support for **custom shaders** alongside **built-in game shaders**
* Replaced manual shader path input with a unified **enum dropdown** and **toggle** to switch between game and custom shaders
* Shader selection is now based on the **game data path set in addon preferences**, making setups portable and shareable
* Variations now use a **prop_search** for easier selection
* **Shader parameters and textures are preserved when switching variations**, as long as the data remains compatible #180
* Parameters that **match their default values are no longer exported**, reducing clutter in output files
* Required Vertex Attributes for selected shader + variation is now visual in the UI
* Parameters are now added dynamically so they now have default, min, max and description (if it exist in the shader)

## FS25 Support (converter and material templates)
* Full material converter from FS19/22 UDIM style vehicle shader to new FS25 material template system
  * Variations
  * Parameters
  * Move UVs under or above Y/V 0 for new wetness feature (interior and windows under and exterior parts above)
  * If no image texture input in BSDF shader Base Color input, it will add image texture with base game "white_diffuse.dds"
  * Cleanup/clear old unused materials
* Brand/Material template selectors for related shader parameters
* Create a new material based on a Material Template and assign it to acitve/selected objects
* New scene property added for defining modDesc.xml, when this is defined it will add any materialTemplates stored in the modDesc.xml to the Brand Material Templates popup (will be shown in the list with "(modDesc)" behind the template name)

![image](https://github.com/user-attachments/assets/5b2dfa1f-cef0-4539-8ce3-98d1c96541d0)
![image](https://github.com/user-attachments/assets/262ac90e-2919-49ea-83da-229609e3b13e)
![image](https://github.com/user-attachments/assets/b25a4538-1900-4522-92a7-936a2e294c2a)


Custom Shader Folders:
Scene properties:
![image](https://github.com/user-attachments/assets/a5094c6a-1243-417a-a68b-3ba3661afd6f)
Enable "Use Custom Shaders" in material properties:
![image](https://github.com/user-attachments/assets/7edea8b1-3351-4e2b-ad46-0c3de7086ebc)
And the shader enum will now use the custom shaders instead:
![image](https://github.com/user-attachments/assets/5424cf18-8c2c-4c0d-81a5-1aa3af8ed86e)

NMC-TBone committed:
Changes to how you add shaders in the UI (#182)

* fix(shader): Prevent resetting parameters in shader settings

Introduce an early return check for the shader's variation. If the selected variation is the same as the previously active one, the function exits early, preventing any unnecessary resetting of shader parameters.

Fix #180

* feat(material): New way of adding shaders

- The shaders will now be automatically loaded in a EnumProperty through the data path set in addon preferences
- This includes a operator that will migrate the old "source" property over to the new "shader" enum property (and preserve all the other shader related data)
- Made it possible to add "material_name" to the load_shader and load_shader_variation operators so it can be called through API without needing to have a active object with a active material to run the operators

* Added support to export new shader setup

StjerneIdioten committed:
Convert from old format at load time

Convert from the old format while loading a file instead through a manual click on an operator

StjerneIdioten committed:
Delete backup functions for old parameter

Delete all the functions that save out the data as it is no longer needed

StjerneIdioten committed:
Fix parameters being overwritten on first load

StjerneIdioten committed:
Compact the UI design

StjerneIdioten committed:
Cleanup shader loading

StjerneIdioten committed:
Introduce new structure for storing shader metadata

StjerneIdioten committed:
Add last bit of changes to keep around for future reference design of this feature

NMC-TBone committed:
Merge branch 'StjerneIdioten:master' into feat/new-shader-picker

NMC-TBone committed:
Merge branch 'develop' into feat/new-shader-picker

NMC-TBone committed:
cleanup after dev merge

NMC-TBone committed:
implemented variations

NMC-TBone committed:
added parameters

NMC-TBone committed:
fixed parameters layout alignment

NMC-TBone committed:
all parameters should now be supported

NMC-TBone committed:
add support for shader textures and update handling in ShaderManager

NMC-TBone committed:
add support for required vertex attributes in ShaderManager and UI

NMC-TBone committed:
handle cases when shader and variation is set through api

NMC-TBone committed:
implemented handling for all old shader properties

NMC-TBone committed:
added export support for new shader setup

NMC-TBone commented on addon/i3dio/ui/shader_picker.py:
```diff
@@ -11,276 +11,212 @@
+    def shader_getter(self):
+        return self.get('shader', 0)
+
+    shader: EnumProperty(
```
Would it make sense to do the same as I did for variations for the shaders as well? (search_prop)
![image](https://github.com/user-attachments/assets/38d652f6-93e0-4497-b715-759b3059cdfc)

NMC-TBone replied:
Would it make sense to do the same as I did for variations for the shaders as well? (search_prop)
![image](https://github.com/user-attachments/assets/38d652f6-93e0-4497-b715-759b3059cdfc)

NMC-TBone reviewed (commented):
I am not really sure what to look after anymore haha.

The only thing that I have not included is support for custom shaders, not entirly sure where or how it should be done. In addition to that, probably 99.9% of cases in game shaders is used 😅Of course thats no reason not to include it. But yea... 🙈

LKAMinco commented:
> I am not really sure what to look after anymore haha.
> 
> The only thing that I have not included is support for custom shaders, not entirly sure where or how it should be done. In addition to that, probably 99.9% of cases in game shaders is used 😅Of course thats no reason not to include it. But yea... 🙈

Maybe add to shader enum Custom Path entry and when it is selected, the path getter will show.

NMC-TBone commented:
> Maybe add to shader enum Custom Path entry and when it is selected, the path getter will show. + if u will not rename the variable and it still will be i3d_attributes.shader_picker.source, then even old blend files will be supported, because value is stored even when is not visible,
> 


Hmm yea adding "Custom Path" entry to enum is possible, but I am not 100% sure what I think about it tbh. Might become hard to find your own "custom shader" in a list with already 50 shaders in it.
I saw @StjerneIdioten added `shader_extra_paths` CollectionProperty to the UI properties, which I guess is a place for users to add custom shaders paths. So in my head it might make more sense to have some kind of toggle/boolean property in the material panel to choose if you want to use game shaders or your own shaders from `shader_extra_paths` collection prop.

> for the shader enum, i recommend to parse all .xml files in data/shaders file, in case Giants will add / remove some shaders.

`populate_shader_cache` will run each time you open a blender file, so that should already be covered I think 🤔

LKAMinco commented:
> > Maybe add to shader enum Custom Path entry and when it is selected, the path getter will show. + if u will not rename the variable and it still will be i3d_attributes.shader_picker.source, then even old blend files will be supported, because value is stored even when is not visible,
> 
> Hmm yea adding "Custom Path" entry to enum is possible, but I am not 100% sure what I think about it tbh. Might become hard to find your own "custom shader" in a list with already 50 shaders in it. I saw @StjerneIdioten added `shader_extra_paths` CollectionProperty to the UI properties, which I guess is a place for users to add custom shaders paths. So in my head it might make more sense to have some kind of toggle/boolean property in the material panel to choose if you want to use game shaders or your own shaders from `shader_extra_paths` collection prop.
> 
> > for the shader enum, i recommend to parse all .xml files in data/shaders file, in case Giants will add / remove some shaders.
> 
> `populate_shader_cache` will run each time you open a blender file, so that should already be covered I think 🤔

You can make shader selection as searchbar, or you can manually place at second index "Custom Path", if the default is None. So it will be always on top of the enum.

NMC-TBone commented:
For the parameters, is there any way we could add default values, min/max values & for FS25 description to each of the parameter? 🤔

Have looked a little bit at it and it seems to not be doable through PropertyGroup properties at least ad FloatVectorProperty only support single float input for min/max anyways.

NMC-TBone committed:
fix issue with param_type

NMC-TBone committed:
add shader cache population on FS Data Folder update

NMC-TBone committed:
fix what might have been a merge conflict

NMC-TBone committed:
set default shader variation name and ensure population on empty collection

NMC-TBone committed:
Merge branch 'develop' into feat/new-shader-picker

NMC-TBone committed:
revert some unecessarry changes to help with conflicts

NMC-TBone committed:
same again

NMC-TBone committed:
Merge branch 'develop' into feat/new-shader-picker

NMC-TBone committed:
filter out None shaders when populating shader enum items

The original problem with parsing should be fixed with #284, but best to be 100% sure anyways.

NMC-TBone committed:
fixed spelling mistake

NMC-TBone committed:
improve descriptions and refine variation handling logic

NMC-TBone committed:
Merge branch 'develop' into feat/new-shader-picker

NMC-TBone committed:
Merge branch 'develop' into feat/new-shader-picker

NMC-TBone committed:
Merge branch 'develop' into feat/new-shader-picker

NMC-TBone committed:
Merge branch 'develop' into feat/new-shader-picker

NMC-TBone committed:
prevent error when creating materials

NMC-TBone committed:
improve layout alignment for shader parameters

NMC-TBone committed:
add functionality to manage custom shader folders

NMC-TBone committed:
add support for custom shader folders and update shader loading logic

NMC-TBone committed:
update shader export logic to support custom shaders

NMC-TBone committed:
add persistent decorator to shader cache handler to ensure it gets loaded properly

NMC-TBone committed:
refactor shader loading logic to iterate through all scenes and improve handling of old shader formats

NMC-TBone committed:
remember parameters and textures when changing variation and set previous value if possible

NMC-TBone committed:
improvements & fixed export with new setup & export params that are not default

NMC-TBone committed:
set correct compare value and removed debug prints

NMC-TBone committed:
iimprove fs_data_path handling and ensure last path is tracked

NMC-TBone committed:
Merge branch 'develop' into feat/new-shader-picker

NMC-TBone committed:
add reset functionality for shader parameters

NMC-TBone committed:
remove depress on reset operator

NMC-TBone committed:
ensure loaded files is shaders by checking root tag (CustomShader)

NMC-TBone committed:
refactor shader parameter handling to use updated ShaderParameter class and improve type consistency

NMC-TBone committed:
add template support for shader parameters and textures; enhance brand color handling in UI

NMC-TBone committed:
restrict brand color display to Farming Simulator 25

NMC-TBone committed:
refactor shader parameter handling to simplify type management and improve default value assignment

NMC-TBone committed:
update brand color display logic to show only when shader parameters include brand color template

Found at that on different platforms the game path can be different e.g. on steam game folder is called "Farming Simulator 25" and from their own store its "Farming Simulator 2025"

NMC-TBone committed:
refactor shader parameter handling to use dynamic parameters

NMC-TBone committed:
enhance shader migration process with version detection and compatibility check

NMC-TBone committed:
enhance shader migration compatibility check for vehicleShader

NMC-TBone committed:
small changes

NMC-TBone committed:
enable custom shaders property based on shader folder availability

NMC-TBone committed:
cleanup

NMC-TBone committed:
feat(material): material templates and FS25 conversion logic

NMC-TBone committed:
enhance material template integration in shader picker

NMC-TBone committed:
always add brandColor as the first panel

NMC-TBone committed:
Merge branch 'feat/new-shader-picker' into feat/fs22-to-25-mat-converter

NMC-TBone committed:
remove unused properties from I3DMaterialTemplates

NMC-TBone committed:
add template search popup operator for material selection

NMC-TBone committed:
improve alignment of shader parameters in the shader group panel

NMC-TBone committed:
Merge branch 'feat/new-shader-picker' into feat/fs22-to-25-mat-converter

NMC-TBone committed:
streamline material template handling and improve shader picker integration

NMC-TBone committed:
add new get_fs_data_path function

NMC-TBone committed:
Merge branch 'feat/new-shader-picker' into feat/fs22-to-25-mat-converter

NMC-TBone committed:
update template parsing to use get_fs_data_path

NMC-TBone committed:
improved clarity

NMC-TBone committed:
add operator to create material from template with assignment options

NMC-TBone committed:
refactor template_to_material and I3D_IO_OT_template_search_popup to enhance parameter handling and descriptions

NMC-TBone committed:
more info

NMC-TBone committed:
add create material from template operator to shader picker panel

NMC-TBone committed:
progress on udim converter

NMC-TBone committed:
cleanup

NMC-TBone committed:
made sure shader is not reset in mats when game version is changed

NMC-TBone committed:
update shader settings export logic and clean up unused shader properties

NMC-TBone committed:
progress

NMC-TBone committed:
change return type of detect_fs_version function from int to str

NMC-TBone committed:
progress

NMC-TBone committed:
remove unused SHADER_PARAMETER_MAX_DECIMALS constant

NMC-TBone committed:
cleanup

NMC-TBone committed:
Merge branch 'feat/new-shader-picker' into feat/fs22-to-25-mat-converter

NMC-TBone committed:
fix merge errors

NMC-TBone committed:
update vehicle shader check and enhance material processing for game version 25

NMC-TBone committed:
streamline shader imports and update shader settings checks

NMC-TBone committed:
enhance UDIM to material conversion for vehicle shaders, supporting negative Y indices and template mapping

NMC-TBone committed:
migration logic fixes

NMC-TBone committed:
wip

NMC-TBone committed:
completed most of 22->25 conversion logic all cases not tested yet

NMC-TBone committed:
refactor

NMC-TBone committed:
remove existing material template previews before creating new ones

NMC-TBone committed:
Add template ID and name handling in material and brand template parsing

NMC-TBone committed:
Refactor helper functions and update material template conversion logic for better clarity and compatibility with FS25

NMC-TBone committed:
completed fs22 to fs25 material converter

NMC-TBone committed:
fix(shape): Ensure correct material order by using mesh slot list

The previous logic for processing mesh triangles collected used materials into an unordered Python `set`. This caused the final list of material IDs and subsets to be generated in an arbitrary order, leading to material mix-ups in the exported file.

This change refactors the logic to use the mesh's own material slot list (`mesh.materials`) as the authoritative source of order. It now iterates through the slots and builds a final, ordered list of *used* materials. All subsequent processing is now based on this correctly ordered list.

This resolves issues where models would export with incorrect materials and also provides a minor performance improvement due to more direct data access.

NMC-TBone committed:
Remove unnecessary mesh update calls after material assignment

NMC-TBone committed:
add support for loading brand material templates from modDesc.xml

NMC-TBone committed:
reset data path to empty string if it was not found

NMC-TBone committed:
make sure default material is added only when needed
